### ISSUE TITLE
Forgone caching in cycles caused much overflow in trait solving

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#61960]
We believe PR #60444 injected a number of regressions of the form "overflow evaluating requirement [...]".

(PR #60444 also injected a compile-time performance regression; that arguably distinct issue is tracked in issue #60846.)

This issue is meant to collect the cases where there was some sort of "overflow evaluating requirement" that was injected by PR #60444. (It is possible that many (or even all) such cases are resolved by follow-up PR #61754. I'll be beta-nominating that PR on its own so that we can evaluate that question in the compiler-team meeting.)

**Comment by pnkfelix** [comment]
In case its not clear from my description: It is *possible* that this issue is entirely resolved by PR #61754, at least on the nightly channel. This umbrella issue is meant to be a place where I can make a list of all the relevant regressions and evaluate whether PR #61754 fixed them, which may inform a future decision on whether or not to beta-backport that PR.

**Comment by matklad** [comment]
I've tried `rustc 1.37.0-nightly (04a3dd8a8 2019-06-18)` and it still takes unreasonable time to compile rust-analyzer ([this commit](https://github.com/rust-analyzer/rust-analyzer/commit/3fa302f3f03e81d17b993c5080f21d545eef0e5b) specifically). I don't know if the compilation succeeds or fails: it has been going on for 10 minutes already.

**Comment by matklad** [comment]
It overflowed half an hour later

**Comment by pnkfelix** [comment]
Note: The compile-time performance regression was/is meant to be tracked by issue #60846. (And maybe we need to open a fresh such issue, since #60846 was closed by PR #61754 )
 
This issue is just meant to track "overflow evaluating requirement" diagnostics that we believe were injected by #60444.

**Comment by pnkfelix** [comment]
triage: P-high, removing nomination.

**Comment by pnkfelix** [comment]
Unassigning self; I still think this is important, but I'm not going to be able to address it, at least not until I return from leave in September.

**Comment by nagisa** [comment]
Pre-triage visit: Doubt this will be looked at before vacation spree ends, but if you want to give it a shot, by all means do so.

**Comment by pnkfelix** [comment]
visiting for triage. Downgrading to P-medium and re-assigning back to self.
